module/zstd/zstd.c

```diff
@@ -0,0 +1,508 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2016-2018 by Klara Systems Inc.
+ * Copyright (c) 2016-2018 Allan Jude <allanjude@freebsd.org>.
+ */
+
+#include <sys/param.h>
+#include <sys/sysmacros.h>
+#include <sys/zfs_context.h>
+#include <sys/zio_compress.h>
+#include <sys/spa.h>
+
+#define	ZSTD_STATIC_LINKING_ONLY
+#include <sys/zstd/zstd.h>
+#include <sys/zstd/zstd_errors.h>
+#include <sys/zstd/error_private.h>
+
+
+#define	ZSTD_KMEM_MAGIC		0x20160831
+
+/* for BSD compat */
+#define	__unused			__attribute__((unused))
+
+
+#ifdef _KERNEL
+#include <linux/sort.h>
+#define	zstd_qsort(a, n, es, cmp) sort(a, n, es, cmp, NULL)
+#else
+#define	zstd_qsort qsort
+#endif
+
+static size_t real_zstd_compress(const char *source, char *dest, int isize,
+    int osize, int level);
+static size_t real_zstd_decompress(const char *source, char *dest, int isize,
+    int maxosize);
+
+void *zstd_alloc(void *opaque, size_t size);
+void zstd_free(void *opaque, void *ptr);
+
+static const ZSTD_customMem zstd_malloc = {
+	zstd_alloc,
+	zstd_free,
+	NULL,
+};
+
+enum zstd_kmem_type {
+	ZSTD_KMEM_UNKNOWN = 0,
+	ZSTD_KMEM_CCTX,
+	ZSTD_KMEM_WRKSPC_4K_MIN,
+	ZSTD_KMEM_WRKSPC_4K_DEF,
+	ZSTD_KMEM_WRKSPC_4K_MAX,
+	ZSTD_KMEM_WRKSPC_16K_MIN,
+	ZSTD_KMEM_WRKSPC_16K_DEF,
+	ZSTD_KMEM_WRKSPC_16K_MAX,
+	ZSTD_KMEM_WRKSPC_128K_MIN,
+	ZSTD_KMEM_WRKSPC_128K_DEF,
+	ZSTD_KMEM_WRKSPC_128K_MAX,
+	/* SPA_MAXBLOCKSIZE */
```

Comment:
Could you put `SPA_OLD_MAXBLOCKSIZE` and `SPA_MAXBLOCKSIZE` after their respective entries.  I found this confusing initially.

Reply:
all entries need to be replaced. they are all confusing and nowhere used. will do that

Reply:
correction. cannot be relaced. SPA_MAXBLOCKSIZE etc have defined values like 1<<24. this will change the enum behaviour. these names have just index behaviour. but i will change it a little bit to make it clear